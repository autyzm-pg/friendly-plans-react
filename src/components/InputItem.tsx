import React from 'react';
import { StyleSheet, TextInput as BaseTextInput, TextInputProps, View } from 'react-native';

import { palette, typography } from 'styles';
import { IconButton } from './IconButton';
import { StyledText } from './StyledText';

interface Props extends TextInputProps {
  label?: string;
  error?: string;
  touched?: boolean;
  styleContainer?: object;
}

interface State {
  isSecureTextVisible: boolean;
}

export class InputItem extends React.PureComponent<Props, State> {
  state = {
    isSecureTextVisible: false,
  };

  toggleSecureTextVisibility = () => {
    this.setState({ isSecureTextVisible: !this.state.isSecureTextVisible });
  };

  renderSecureTextSwitch = () => {
    const { secureTextEntry } = this.props;
    const { isSecureTextVisible } = this.state;
    if (!secureTextEntry) {
      return;
    }
    return (
      <IconButton
        name={isSecureTextVisible ? 'eye-off-outline' : 'eye-outline'}
        onPress={this.toggleSecureTextVisibility}
        containerStyle={styles.secureTextIconContainer}
      />
    );
  };

  render() {
    const { label, error, touched, style, secureTextEntry, styleContainer } = this.props;
    return (
      <View>
        {!!label && <StyledText style={styles.label}>{label}</StyledText>}
        <View style={[styles.inputContainer, styleContainer]}>
          <BaseTextInput
            {...this.props}
            style={[styles.input, style]}
            secureTextEntry={secureTextEntry && !this.state.isSecureTextVisible}
          />
          {this.renderSecureTextSwitch()}
        </View>
        {!!error && (
          <View style={styles.errorContainer}>{touched && <StyledText style={styles.error}>{error}</StyledText>}</View>
        )}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  label: {
    ...typography.caption,
    color: palette.textDisabled,
    marginBottom: 4,
  },
  inputContainer: {
    height: 44,
    borderRadius: 4,
    borderColor: palette.border,
    borderWidth: 1,
  },
  input: {
    height: 42,
    justifyContent: 'center',
    paddingHorizontal: 16,
    ...typography.subtitle,
    color: palette.textBlack,
  },
  errorContainer: {
    height: 26,
    justifyContent: 'center',
  },
  error: {
    ...typography.caption,
    color: palette.error,
  },
  secureTextIconContainer: {
    position: 'absolute',
    right: 0,
    top: 0,
    height: 42,
    width: 60,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
