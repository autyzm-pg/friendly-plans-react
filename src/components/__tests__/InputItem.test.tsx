import { shallow } from 'enzyme';
import React from 'react';

import { InputItem } from '../InputItem';

describe('InputItem', () => {
  const props = {
    label: 'input label',
  };

  it('should match snapshot', () => {
    const wrapper = shallow(<InputItem {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render with secure text entry', () => {
    const wrapper = shallow(<InputItem {...props} secureTextEntry />);
    expect(wrapper.find('IconButton')).toExist();
  });

  it('should toggle secure text visibility', () => {
    const wrapper = shallow(<InputItem {...props} secureTextEntry />);
    const instance = wrapper.instance() as InputItem;
    expect(instance.state.isSecureTextVisible).toBe(false);
    instance.toggleSecureTextVisibility();
    expect(instance.state.isSecureTextVisible).toBe(true);
  });

  it('should render with error visible if error present and touched', () => {
    const error = 'oh my';
    const wrapper = shallow(<InputItem {...props} error={error} touched />);
    expect(wrapper.find(`StyledText[children="${error}"]`)).toExist();
  });
});
