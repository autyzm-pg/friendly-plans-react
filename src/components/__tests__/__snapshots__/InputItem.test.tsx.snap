// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputItem should match snapshot 1`] = `
<View>
  <StyledText
    style={
      Object {
        "color": "#9B9B9B",
        "fontFamily": "Rubik-Medium",
        "fontSize": 12,
        "letterSpacing": 0,
        "marginBottom": 4,
      }
    }
  >
    input label
  </StyledText>
  <View
    style={
      Object {
        "borderColor": "#dedede",
        "borderRadius": 4,
        "borderWidth": 1,
        "height": 44,
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      label="input label"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "color": "#16107B",
            "fontFamily": "Rubik-Regular",
            "fontSize": 16,
            "height": 42,
            "justifyContent": "center",
            "letterSpacing": 0,
            "paddingHorizontal": 16,
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "height": 26,
        "justifyContent": "center",
      }
    }
  />
</View>
`;
